### PR TITLE
Feature/many to many features and account type

### PR DESCRIPTION
**Pull request by ARBaynes**
Hi @gtvj, here's the pull request for the work I've done on the AccountType model (linking it to both features and users).

### CODE REVIEW

**Comment by ARBaynes** [comment]
Hi @gtvj, I've fixed the conflict (I think it arose after the SubscriptionController was changed by my other pr), so hopefully it should be okay for review now!

**Comment by gtvj** [comment]
Thanks @ARBaynes - I'll look at it this afternoon.

**Comment by gtvj** [comment]
Hi @ARBaynes - I haven't finished looking at this yet but have come across a problem. I've tried registering a few users and it doesn't seem that they are being associated with an account type. I've been trying to establish why but haven't got anywhere yet. I'll take a look tomorrow.

File: database/seeds/AccountTypesTableSeeder.php [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+class AccountTypesTableSeeder extends Seeder
+{
+    /**
+     * Run the database seeds.
+     *
+     * @return void
+     */
+    public function run()
+    {
+        DB::table('account_types')->insert([
```
Comment (gtvj) [inline database/seeds/AccountTypesTableSeeder.php]: These two inserts can be achieved with a single call:

```php
        DB::table('account_types')->insert([
            ['name' => 'Free', 'description' => 'A free account'],
            ['name' => 'Subscriber', 'description' => 'A subscriber account']
        ]);
```

File: app/Http/Controllers/ActivationController.php [hunk 2]
```diff
@@ -25,6 +25,12 @@ public function verify_email()
         return view('verify_email', compact('content', 'user'));
     }
 
+    public function send_email ()
```
Comment (gtvj) [inline app/Http/Controllers/ActivationController.php]: Hi @ARBaynes - do the introduction of `send_email()` and refactoring of `resend_activation_code()` relate to the introduction of the many-to-many relationship? If not, I think it might be best if these changes are extracted to a separate commit and PR.

File: app/Http/Controllers/Auth/RegisterController.php [hunk 3]
```diff
@@ -71,11 +75,14 @@ protected function create(array $data)
             'email_confirmation' => $data['email_confirmation'],
             'password' => Hash::make($data['password']),
             'verification_code' => strtoupper(Str::random(4)),
-            'email_verified' => false,
-            'account_type_selection' => $data['account_type_selection']
+            'email_verified' => false
         ]);
 
-        \Mail::to($user)->send(new VerificationEmail($user));
+        $user->account_type()->associate(
```
Comment (gtvj) [inline app/Http/Controllers/Auth/RegisterController.php]: Hi @ARBaynes - For some reason, this doesn't seem to be resulting in the relationship being created. While looking into this I noticed that `User` has a foreign key for `account_type_id` which suggests a one-to-many relationship. Because users' account types will change as they subscribe, unsubscribe and cancel a subscription mid payment cycle, the modelling is a bit more complex. It was because of circumstances like this that I was so keen for us to have a diagram of the data model included in the repository. It's much easier to make the changes to a diagram, consider whether those changes will meet our needs and then make the changes to code (including the updated model in the PR).

File: app/User.php [hunk 4]
```diff
@@ -34,4 +34,14 @@ public function research_interests()
     {
         return $this->belongsToMany('App\ResearchArea');
     }
+
+    /**
+     * Get the account type associated with the user.
+     */
+    public function account_type()
```
Comment (gtvj) [inline app/User.php]: Hi @ARBaynes - do we need to be using `withDefault()` here? At the point we associate a user with an account type, we know what type of account they have. It took me a while to get my head round what was going on here and it seems to be making things more complicated than necessary for a `belongsTo()` relationship

File: database/migrations/2014_10_12_000000_create_users_table.php [hunk 5]
```diff
@@ -19,11 +19,14 @@ public function up()
             $table->string('email')->unique();
             $table->string('email_confirmation');
             $table->string('password');
-            $table->string('account_type_selection')->nullable();
             $table->string('verification_code')->nullable();
             $table->boolean('email_verified');
+            $table->integer('account_type_id')->nullable();
```
Comment (gtvj) [inline database/migrations/2014_10_12_000000_create_users_table.php]: Hi @ARBaynes - This isn't a change you've introduced here looking at the code I'm not sure `account_type_id` needs to be nullable. It permits us to create a user without associating an account type. Ultimately, we'll need a many-to-many here but it's just a thought. What do you think?

File: resources/views/dashboard.blade.php [hunk 6]
```diff
@@ -9,40 +9,32 @@
...
                     <div class="col-md-4">
                         <article>
                             <div class="entry-header">
                                 <h2>{{ $user->name }}</h2>
                             </div>
-
                             <div class="entry-content">
-                                <h3>{{$content['headings']['account_type_h3']}}</h3>
-                                {{$content['example_text']}}
+                                <h3>{{$user->account_type()->first()->name}}</h3>
```
Comment (gtvj) [inline resources/views/dashboard.blade.php]: Because of the issue with `account_type()` not being created when a user is registered I'm getting an error on the dashboard page. This is a good reason not to permit `account_type_id` to be `nullable()` in the migration that creates the users table.

File: resources/views/thank_you.blade.php [hunk 7]
```diff
@@ -21,7 +21,7 @@
                                 <p class="pull-right">
                                    <a class="button" href="https://discovery.nationalarchives.gov.uk">{{$content['go_to_discovery']}}</a>
 
-                                   <a class="button" href="{{ route('dashboard') }}">{{$content['go_to_dashboard']}}</a>
+                                   <a class="button" href="{{ route('your-research-interests') }}">{{$content['go_to_dashboard']}}</a>
```
Comment (gtvj) [inline resources/views/thank_you.blade.php]: Hi @ARBaynes - this is a minor point but does this change relate to the PR?

**Comment by gtvj** [comment]
I’ve added a number of specific comments but also wanted to provide you with an overview of my observations since I know you’d like to use this PR as a learning experience. The main feedback I’d give is that:

* Some changes seem unrelated to the PR. This presents two problems: 
    * It’s really hard for reviewers to get their head round the code for complex PRs, and it gets really confusing if there are changes that might be unrelated
    * Where other work is mixed in with a commit we lose the ability to see the story of our code and, more importantly, to use some advanced features of Git that can save us work in the future (i.e. completely undoing the work in a commit and replacing it with something else) 
* The users aren’t being associated with an account type on registration
* It seems the data model might not be right (we need a many-to-many relationship between users and account types)

**Recommended next steps:**

With complex changes such as these it really helps to take what might seem like an overly structured approach.

At this stage I would close this PR (but keep it on GitHub as a reference) and do the work in a new branch taken from develop. This will be much more straightforward than introducing new commits to unpick and reimplement what’s there. My recommendation would be to follow this sequence: 

* Update the model diagram and validate it against what we know. Once that’s done and we’re sufficiently confident we can code against it, I'd include the updated diagram in the repository (as `model_diagram_v_0_2.png`, updating the `README.md` accordingly)
* Introduce the `AccountType` model
* Create a many-to-many relationship between `AccountType` and `Feature`, along with associated seeds
* Create a many-to-many relationship between `AccountType` and `User`, creating/updating associated seeds
* Update `RegisterController` and associated views etc. so that the models are populated on registration

I’d be very happy to pair with you (or help in any other way) on this or, if you’d prefer, I’d be happy to make these changes myself.
